WebServices/AzureNotificationHub/NotificationHubSample/NotificationHubSample/MainPage.xaml.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace NotificationHubSample
+{
+    [DesignTimeVisible(true)]
+    public partial class MainPage : ContentPage
+    {
+        public MainPage()
+        {
+            InitializeComponent();
+        }
+
+        protected override void OnAppearing()
+        {
+            base.OnAppearing();
+        }
+
+        protected override void OnDisappearing()
+        {
+            base.OnDisappearing();
+        }
+
+        public void AddMessage(string message)
+        {
+            Device.BeginInvokeOnMainThread(() =>
```

Comment:
Not yet seen where this is invoked from, but does it require wrapping in `BeginInvokeOnMainThread`?

Reply:
Yes

OnMessageReceived happens on a background thread on Android. Messages may also come in on a background thread on iOS, I didn't notice as I already had this set up to force UI thread execution.